### DISCUSSION TITLE
Derivative Explanation

### DISCUSSION THREAD

**Discussion by tloom** [pyvista/pyvista#2988] [Q&A]
I have had quite a few more attempts to understand how derivatives are calculated since #2680 and have had little success.

Would I be able to get an insight into how these are calculated? 

I have included two test cases where I get different answers for the derivatives based on whether the data is StructuredGrid or PolyData. 

```py
import numpy as np
import pyvista as pv
def func(x,y,z):
    gridst=pv.StructuredGrid(x,y,z)
    gridst['z']=z.flatten()
    gridply=gridst.extract_surface() #polydata object
    gridst=gridst.compute_derivative('z')
    gridply=gridply.compute_derivative('z')
    return gridst['gradient'],gridply['gradient']

n=10
"""Case 1, plane inclined by 45 deg in yz plane"""
x_=np.linspace(0,1,n)
y_=np.linspace(0,1,n)

x, y= np.meshgrid(x_, y_, indexing='ij')
z=y

gst,gply=func(x,y,z)
print('Case1:')
print('Structured gradient, '+str(gst[0]))
print('Polydata gradient, '+ str(gply[0])) #as all gradients in plane are equal

"""Case 2, plane inclined by 45 deg in xy plane"""
x_=np.linspace(0,1,n)
z_=np.linspace(0,1,n)

x2, z2= np.meshgrid(x_, z_, indexing='ij')
y2=x2

gst2,gply2=func(x2,y2,z2)
print('Case2:')
print('Structured gradient, '+str(gst2[0]))
print('Polydata gradient, '+ str(gply2[0]))
```

**Comment by MatthewFlamm** [comment]
The gradient is calculated using `vtkGradientFilter` https://vtk.org/doc/nightly/html/classvtkGradientFilter.html

I suspect there is some weirdness going on with the degeneracy in the structured grid in the 2D case.  I'm not sure if this is a bug in vtk, pyvista or a 'feature'.

**Comment by tloom** [comment]
Ah okay, I was looking for a description of the mathematics that is used in vtkGradientFilter and why the method apparently changes between structured data and polydata if you might be able to direct me?

**Comment by RichardScottOZ** [comment]
https://discourse.paraview.org/t/numerics-of-gradient-of-unstructured-dataset-filter/638/5
